input refactoring: create input_driver.c

markwkidd opened the pull request:
This PR, which compiles and runs, is intended foremost for @jdgleaver 

As one would expect, the process of refactoring input_drivers into a more modular form touches a lot of code. I have submitted this PR now, even though it obviously does not pull out all of the input_driver.h functions, because I hope it goes far enough to get your feedback on my approach.

As you can see, I have begun implementing a new `input_driver_state` struct within `input_drivers` where the relevant fields can be moved out of the `ra_state` struct in `retroarch_data.h`. Generally speaking, I'm trying make sure that there are `get` and `set` functions available for the fields in the new `input_driver_state` so that, once complete, there won't need to be any direct external access to the new struct.

At this point I am open to any feedback.

My personal preference would be to break the overall refactor process into several incremental PRs that can be reviewed and merged at a scale where they can be more easily read. But that too is open to feedback. Thanks!

markwkidd commented:
@jdgleaver I'll add that the current captures the process part-way through my 'second wave' of migrating to `get_input_driver()` and `input_driver_get_data()` in retroarch.c. The first wave was exposing the new input state struct fields directly to retroarch.c in order to get them moved to the new struct beforehand.

twinaphex commented:
I will leave this open for @jdgleaver  to comment on, but I already see some things we would want to avoid -

'input_driver_t *current_input           = get_input_driver();
   void *input_data                        = input_driver_get_data();'

All these small getter functions kinda ruined our code design previously with earlier refactors, and it's probably a mistake to fall back on that again. As part of this refactoring, we also have to consider how we lay out the data and what kind of data should be together in the same struct so we don't have to do all this pointer grabbing, or reduce it at least to a minimum.

What do you think, @jdgleaver ? Anything to add?

markwkidd commented:
Thanks, this is exactly the kind of feedback I am hoping for by submitting a draft PR at an early stage. I'd like to make sure that I have the foundation right.

jdgleaver commented:
@markwkidd Thanks for making a start on this!

I'm currently in the middle of something difficult and tedious that I need to finish before I get deep into reviewing anything else, but in the meantime I agree with twinaphex. What I've tried to do with the libraries I've added - and what I think we want moving forwards - is the following type of design model:

- A library should have no internal global state at all (unless really necessary - e.g. mutex locking objects often make sense as static globals)
- A function should be provided to create/return a new malloc'd/calloc'd library-specific state object, with another function to free it
- All other functions should accept as an argument a pointer to said library-specific state object

This way the library-specific state can be grabbed/handled/stored externally (in some singular well-managed global state struct), and the actual library code just operates on what you feed it.

I think switching over to this for the input driver code will indeed take several iterations, in multiple PRs :)

But for now, you could probably start by keeping the global state in `ra_state` (probably using your compound `input_driver_state` struct - which should be typdef'd as `input_driver_state_t` BTW :) ), and passing it as the first argument to all your `input_*()` functions.

markwkidd commented:
Thank you @twinaphex and @jdgleaver I believe I have the gist of the pattern you are describing. I know I have seen some code structured that way, so I'll look for that approach in other library functions to have as a reference.

That said, if you can think of any library you would recommend as particularly a good model for me, please let me know.

In the meantime, I'll get back to work on this PR.

jdgleaver commented:
There are quite a few good examples - off the top of my head, these simple libraries follow the kind of pattern we're looking for:

https://github.com/libretro/RetroArch/blob/master/core_backup.h
https://github.com/libretro/RetroArch/blob/master/runtime_file.h

But I think this one is particularly relevant in the current situation:

https://github.com/libretro/RetroArch/blob/master/disk_control_interface.h

Here the actual library object (https://github.com/libretro/RetroArch/blob/f831fe9ad0386b0a81aab5aed7558aa6d8b1dafd/core.h#L61) is kept in an external struct, and there are no init()/free() functions (which is what I think you'll be doing with the input code for at least the first few revisions, until some kind of init()/free() setup can be established)

markwkidd commented:
@jdgleaver thank you for the links, that was very helpful. I have remade the PR according to the approach you described. I understand you may not respond right away. I would appreciate any review you can give when you have time.

When you have a chance,  I do have one specific question as I continue: Is it acceptable for me to add fields to the `settings_t` struct that seem to fit? If you look at the changes for `configuration.h`, I have moved `input_max_users` and `input_axis_threshold` from `rarch_state`.

jdgleaver commented:
@markwkidd Ah, you have been busy! There is a lot to see here - I will do my best to review the code next week :)

> When you have a chance, I do have one specific question as I continue: Is it acceptable for me to add fields to the settings_t struct that seem to fit? If you look at the changes for configuration.h, I have moved input_max_users and input_axis_threshold from rarch_state.

Yes, as far as I can tell this is a wholly positive change. I can't really think of a good reason for the existing behaviour, where things like `input_max_users` and `input_axis_threshold` are treated as special cases in `rarch_state`. *Possibly* there is some miniscule performance benefit from avoiding lookups in the `settings_t` struct, but that hardly seems relevant. Moving settings values to the settings struct would appear to be a cleaner and more maintainable approach.

---

One other small thing: where you have functions like `set_rumble_state()` and `set_input_sensor_state()`, I'd prefer these to have an `input` prefix (i.e. `input_set_rumble_state()`, `input_set_sensor_state()`). As a general  rule, we always try to prefix functions with their 'category', so to speak.

markwkidd commented:
Awesome, thanks. I changed those function names to begin with "input" but I'll try to leave it alone until you can take a look.

jdgleaver commented:
@markwkidd Apologies for the delay! I've at last managed to finish my current project (#12749) - I need to take a little break now, but I should be free to review your PR on Monday. I'll let you know as soon as I have some feedback :)

markwkidd commented:
I'm pretty excited about the core option enhancements too :) Thanks, and have a good weekend!

markwkidd commented:
@jdgleaver thank you for helping me with this PR. I have just pushed a commit to not pass the settings_t, but only the specific field needed.

From my understanding, that will make this PR ready to merge, right? 

Is it better for me to squash the commits on this side, or would you all like to handle that from the repo admin side? I made the commits as semantic/granular as possible, and they each should compile, but I don't think there's much value in keeping them individually.

markwkidd commented:
I see other PRs being squashed by the contributor, so I'll go ahead and do that here as well.

jdgleaver commented:
@markwkidd Thanks for the update, and for squashing the commits (that is indeed the correct approach)

> From my understanding, that will make this PR ready to merge, right?

Yes, this all looks good to me and I think the PR is ready to go.

@twinaphex Please feel free to merge this at your leisure :)

twinaphex commented:
Hi, I glanced through the PR, but I didn't see the file being added to Griffin? griffin/griffin.c in specific.

markwkidd commented:
@twinaphex you are right, it was not added to griffin.c -- I was not aware of the role of griffin, but the format makes the necessary change easy to figure out. I've now added `input_driver.c` to `griffin.c`, squashed, and updated this PR.

twinaphex commented on input/input_driver.c:
```diff
@@ -39,16 +39,29 @@ bool input_driver_set_rumble(
 
+   if (!driver_state)
+   {
+      RARCH_ERR("Argument 'driver_state' is NULL in function 'input_driver_set_rumble'\n");
```
I'm not a fan of the kind of warning spam this would cause.

markwkidd replied:
Understood. Fixed!

twinaphex reviewed (changes requested):
I really don't like all the warning spam you added in that last commit. This could seriously harm runtime performance for the set_state functions.

markwkidd commented:
I believe I have found the source of the RS-90 build error, but it's going to take a little bit of refactoring to solve. I'll check back in when I have it worked out.

markwkidd commented:
@jdgleaver when you are able to build this PR for RS-90 again, I think I should have it fixed.

There were three arrays declared as `static` in retroarch_data which I had moved to input_driver.h while retaining the static label. I think the RS-90 codepath 'double included' those static declarations. Now those arrays are more correctly instantiated in input_driver.c with an `extern` in input_driver.h

jdgleaver commented:
@markwkidd Apologies for the delay, and thank you for the fix. Making those arrays `extern` is indeed the correct approach :)

There's still one little build error, though, caused by missing `HAVE_HID` guards. This diff should do the trick:

```diff
diff --git a/input/input_driver.c b/input/input_driver.c
index f659a6ff90..42434b157e 100644
--- a/input/input_driver.c
+++ b/input/input_driver.c
@@ -262,6 +262,7 @@ input_driver_t *input_drivers[] = {
    NULL,
 };
 
+#ifdef HAVE_HID
 hid_driver_t *hid_drivers[] = {
 #if defined(HAVE_BTSTACK)
    &btstack_hid,
@@ -278,6 +279,7 @@ hid_driver_t *hid_drivers[] = {
    &null_hid,
    NULL,
 };
+#endif
 
 /**************************************/
 
diff --git a/input/input_driver.h b/input/input_driver.h
index bebc71fcb8..73f64b7ee2 100644
--- a/input/input_driver.h
+++ b/input/input_driver.h
@@ -721,7 +721,9 @@ void input_config_reset(void);
 
 extern input_device_driver_t *joypad_drivers[];
 extern input_driver_t *input_drivers[];
+#ifdef HAVE_HID
 extern hid_driver_t *hid_drivers[];
+#endif /* HAVE_HID */
 
 extern input_driver_t input_android;
 extern input_driver_t input_sdl;
```

markwkidd commented:
@jdgleaver thank you for spotting the HAVE_HID issue! I have pushed a fix.

jdgleaver commented:
Great, thanks. Looks like we're good to go :)

twinaphex commented:
I'm sorry but open merging this, there were too many build failures on many platforms.

https://git.libretro.com/libretro/RetroArch/-/jobs/753438

We will have to revert this for the time being. We advise you making a new PR where you get all these issues fixed. 

/c/git/libretro-super/retroarch/input/drivers_joypad/qnx_joypad.c:   qnx_input_t *qnx                     = (qnx_input_t*)input_driver_get_data();
/c/git/libretro-super/retroarch/input/drivers_joypad/qnx_joypad.c:   qnx_input_t            *qnx    = (qnx_input_t*)input_driver_get_data();
/c/git/libretro-super/retroarch/input/drivers_joypad/qnx_joypad.c:   qnx_input_t            *qnx    = (qnx_input_t*)input_driver_get_data();

This function was entirely missing and is needed for iOS and other platforms.

c:\gitlab-runner\builds\-sukb9sf\2\libretro\retroarch\griffin\../input/input_driver.c(24) : fatal error C1083: Cannot open include file: 'retroarch.h': No such file or directory

Plenty of wrong header includes - you're supposed to do relative header path includes, like '../retroarch.h' inside the input/ folder.

schellingb commented:
When compiling with Visual Studio (which uses griffin) I get the following compiler/linker errors with this.
```
1>griffin/../input/input_driver.c(24): fatal error C1083: Cannot open include file: 'retroarch.h': No such file or directory
1>griffin/../input/input_driver.c(25): fatal error C1083: Cannot open include file: 'verbosity.h': No such file or directory
1>griffin/../input/input_driver.c(26): fatal error C1083: Cannot open include file: 'configuration.h': No such file or directory
1>griffin.obj : error LNK2001: unresolved external symbol input_builtin_autoconfs
```

I had to change the top of input_driver.c to
```c
#include "../retroarch.h"
#include "../verbosity.h"
#include "../configuration.h"
```
and add this to griffin.c (after input/input_keymaps.c)
```c
#include "../input/input_autodetect_builtin.c"
```

markwkidd commented:
@twinaphex different platforms must have added different include folders in their build environment, because I was able to compile without specifying `../verbosity.h` etc in msys, and others have been able to compile this PR in their environments too.

Is there any way to check and standardize which include paths are provided for which platform? Other than that, I can't think of any way to keep up with platform-specific differences which are hard to compile for.

I notice that LGTM doesn't catch these kinds of platform specific code where `#define` is used

markwkidd commented:
The divergence in the griffin and Makefile approaches to `-I` include folders looks to be recent. Here's where I think the difference arises: https://github.com/libretro/RetroArch/commit/28541fe8267c6407176cdb85f063d1eff20c00ca

schellingb commented:
In general my belief is that the include path list should be kept as small as possible. This automatically means that a bunch of source code files that are always distributed in one unit (like this repository) should always refer to each other with relative paths.

Though yet the include path in my Visual Studio project is
- gfx\include\dxsdk
- gfx\include
- libretro-common\include
- libretro-common\include\compat\zlib
- deps
- deps\glslang
- deps\glslang\glslang
- deps\rcheevos\include
- deps\SPIRV-Cross

Some of those are probably not feasible to remove due to not wanting to touch external code bases (things in deps). The rest being required seems just a result of laziness (or not caring).

So to make things short, I'd prefer if . was removed from the Makefile include list.

markwkidd commented:
@schellingb that makes sense to me. I suppose I should try to see if there's another way to accomplish the purpose of the commit I linked so I can submit a PR to remove `.`

Hopefully the wildcard wasn't added to get around something particularly tiresome to do the right way

schellingb commented:
Yeah, it's hard to know if something or someone now doesn't rely on that `.` being there. It shouldn't have been added in the first place.

markwkidd commented:
I've made a PR for the valgrind include path issue: https://github.com/libretro/RetroArch/pull/12820

markwkidd commented:
New and improved PR here: https://github.com/libretro/RetroArch/pull/12823
